### ISSUE TITLE
[integration_test] Needs null safety migration

### ISSUE THREAD

**Issue by creativecreatorormaybenot** [flutter/flutter#71379]
Because many plugins depend on `integration_test` in order to do their testing, the package should be migrated to support null safety.

https://pub.dev/packages/integration_test

**Comment by pedromassangocode** [comment]
This was being worked in https://github.com/flutter/plugins/pull/3158 but according to https://github.com/flutter/plugins/pull/3158#issuecomment-726399499 the package is being moved to flutter repository so not really sure if this was already addressed or not.
Cc @dnfield @ferhatb

**Comment by dnfield** [comment]
This is blocked by migrating package:vm_service.

**Comment by dnfield** [comment]
@bkonyi

**Comment by bkonyi** [comment]
package:vm_service is basically ready to go, but there's a bit of a dance we need to do in order to land. We'll need to also migrate package:dwds and likely DevTools to null safety. @devoncarew, is there anything else I've missed here?

**Comment by devoncarew** [comment]
Those are the bigger users that I'm aware of.

The rollout could probably do with a little planning - perhaps in an issue at flutter/devtools? Specifically:
- package:dwds pins to a very specific vm_service version - https://github.com/dart-lang/webdev/blob/master/dwds/pubspec.yaml#L38
- once package:dwds upgrades, consumers of dwds can then only ship new releases to nnbd capable sdk (1.12.0+)
- that's not an issue for the main consumer of package:dwds - flutter/flutter

DevTools does want to be able to ship to older stable SDKs (2.10 say).The web app has a dep on vm_service (https://github.com/flutter/devtools/blob/master/packages/devtools_app/pubspec.yaml#L41). We can likely develop that against a dev sdk - the compiled output is used by the user, not the devtools_app package itself. The devtools server also has a dep on vm_service (https://github.com/flutter/devtools/blob/master/packages/devtools_server/pubspec.yaml#L26). We likely couldn't move that dep to the null safe one until we no longer need to support 2.10 for new devtools releases.

cc @grouma @kenzieschmoll for dwds and devtools planning here

**Comment by grouma** [comment]
> Those are the bigger users that I'm aware of.
> 
> The rollout could probably do with a little planning - perhaps in an issue at flutter/devtools? Specifically:
> 
> * package:dwds pins to a very specific vm_service version - https://github.com/dart-lang/webdev/blob/master/dwds/pubspec.yaml#L38
> * once package:dwds upgrades, consumers of dwds can then only ship new releases to nnbd capable sdk (1.12.0+)

Although not a deal breaker, this puts `package:webdev` in an odd state. Given how `webdev` is used as a globally activated package, we don't want it to require dev SDKs. If we upgrade `package:dwds`, we may be blocked on publishing new versions of `package:webdev` until the stable SDK lands. However, we should be able to publish new versions of `package:dwds` whenever.

> * that's not an issue for the main consumer of package:dwds - flutter/flutter
> 
> DevTools does want to be able to ship to older stable SDKs (2.10 say).The web app has a dep on vm_service (https://github.com/flutter/devtools/blob/master/packages/devtools_app/pubspec.yaml#L41). We can likely develop that against a dev sdk - the compiled output is used by the user, not the devtools_app package itself. The devtools server also has a dep on vm_service (https://github.com/flutter/devtools/blob/master/packages/devtools_server/pubspec.yaml#L26). We likely couldn't move that dep to the null safe one until we no longer need to support 2.10 for new devtools releases.
> 
> cc @grouma @kenzieschmoll for dwds and devtools planning here

@bkonyi I think it's worth setting up a quick sync to ensure we are all on the same page here.

**Comment by devoncarew** [comment]
To close the loop here, given that several different packages and tools have tight version range deps on package:vm_service, it may be useful if this particular reference to it was in-lined. This file: https://github.com/flutter/plugins/blob/master/packages/integration_test/lib/integration_test.dart seems to make light use of the library, so could just as easily use a small, hand-written wrapper that speaks to the VM via sending and receiving the json RPC calls.

**Comment by dnfield** [comment]
flutter_tools also will want to migrate to null safety at some point, and also depends on vm_service.

integration_test only makes light use, but flutter_driver makes heavier use, and integration test has dependencies on flutter_driver. We could just avoid using vm_service, but we'd basically be re-implementing it there, and it's not a completely trivial task.

**Comment by grouma** [comment]
> flutter_tools also will want to migrate to null safety at some point, and also depends on vm_service.

This shouldn't be on the critical path though? We should be able to wait until the next stable release.

**Comment by dnfield** [comment]
No, flutter_tools is separate from this completely. But flutter_driver is not separate.

What's the timeline for the next stable release that would make all of this easier?

**Comment by dnfield** [comment]
@grouma @bkonyi @devoncarew - to clarify again, in case I haven't made it clear enough: `integration_test` and `flutter_driver` both have to be migrated to close this bug. There is significant use of `vm_service` in `flutter_driver` that I would not feel comfortable rolling into the Flutter repo.

It would really help to have a clearer timeline (in this bug or a doc/sheet or somewhere) that we can follow to know when this gets unblocked and what its current status is. If this really is blocked until the next stable release of Dart, it'd be good to know what the date (or at least the estimated date) is for that.

/cc @csells

**Comment by grouma** [comment]
Thanks for raising your concerns @dnfield. I was under the impression that Devon's suggestion was sufficient. I set up a sync so we can better coordinate the `package:vm_service` migration.

**Comment by goderbauer** [comment]
This is currently blocking https://github.com/flutter/flutter/issues/67447.

**Comment by devoncarew** [comment]
FYI, @bkonyi has migrated package:vm_service to null safety; https://pub.dev/packages/vm_service @ `6.0.0-nullsafety-dev`.

**Comment by ditman** [comment]
This is great news! Thanks @bkonyi!!!

**Comment by dnfield** [comment]
This is blocked on getting package:vm_service rolled into package:coverage and package:test_core.

That is blocked on getting some additional testing done with dwds and dev_tools.

**Comment by dnfield** [comment]
Unassigning since I'm not working on the upstream blocking issues and can't work on this until they're resolved.

**Comment by devoncarew** [comment]
@dnfield - are there issues filed for the blocking issues mentioned above?

**Comment by dnfield** [comment]
@bkonyi @grouma @kenzieschmoll might know. They're working on testing the new vm_service package with dwds and devtools. Once that's done, it has to be rolled into coverage (https://github.com/dart-lang/coverage/issues/321). Then it needs to go into test_core (https://github.com/dart-lang/test/issues/1409).

**Comment by grouma** [comment]
@kenzieschmoll and I have been running into some issues and have been working with @bkonyi to resolve them. Expect some delays.

**Comment by pcsosinski** [comment]
integration_test is no longer part of the app template, so moving milestones

**Comment by ditman** [comment]
@pcsosinski this is widely used to tests web plugin packages; any way of bumping up the priority on this side of the ecosystem? Otherwise we can't migrate web plugins to null-safety (or we lose most of the testing)

**Comment by grouma** [comment]
Update: @bkonyi is addressing https://github.com/dart-lang/sdk/issues/44683 which should unblock us from moving forward with the necessary changes.

**Comment by ditman** [comment]
_/me celebrates_ 🥳

**Comment by grouma** [comment]
`package:coverage` and `package:test` have been updated to support the latest null safety version of `package:vm_service`. There is a blocking issue preventing `package:dwds` from migrating. See https://github.com/dart-lang/sdk/issues/44725. It's an easy fix that @bkonyi is addressing.

**Comment by dnfield** [comment]
There are a few remaining blockers to landing NNBD vm_service in the repo:

- NNBD `package:image` has to be published - or, at least, a version that allows us to transitively get the right version of crypto.
- `webdriver` -> `sse` -> `uuid` needs to be updated to allow us to get the right version of UUID

We might also need to see `sse` itself have a NNBD migration.

**Comment by dnfield** [comment]
Oh, it looks like `sse` does not depend on UUID anymore, but that version has not been published.

**Comment by dnfield** [comment]
vm_service has rolled into the repo now.

I'm starting with a migration of fuchsia_remote_debug_protocol

**Comment by pichillilorenzo** [comment]
So, when is possible to have the `integration_test` plugin with support for null safety? Is there any deadline?

Thanks all of you!

**Comment by dnfield** [comment]
There are still a couple migrations ahead of it. It'd hard to say right now because I'll have to make sure each one doesn't break tests in upstream repositories and may have to revert them if we run into unexpected behavior.

It will take more than a week or two but hopefully less than a month or two :)

**Comment by dnfield** [comment]
This has been going relatively smoothly. All upstream dependencies are migrated now and driver appears to be working. I've opened a patch to migrate integration_test.

**Comment by dnfield** [comment]
Also - I do not plan to publish this migration to pub. It will be included in the flutter SDK. I can re-evaluate that if it's causing a lot of problems.

**Comment by dnfield** [comment]
Reopening for revert.

**Comment by lukepighetti** [comment]
I'm in a null-safe Flutter project and I'm getting warnings that `integration_test` is a legacy library. Is there a solution?


```
% flutter doctor --verbose
[✓] Flutter (Channel beta, 1.25.0-8.3.pre, on macOS 11.1 20C69 darwin-x64, locale en-US)
    • Flutter version 1.25.0-8.3.pre at /Users/lukepighetti/development/flutter
    • Framework revision 5d36f2e7f5 (3 weeks ago), 2021-01-14 15:57:49 -0800
    • Engine revision 7a8f8ca02c
    • Dart version 2.12.0 (build 2.12.0-133.7.beta)

[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/lukepighetti/Library/Android/sdk
    • Platform android-30, build-tools 29.0.3
    • ANDROID_HOME = /Users/lukepighetti/Library/Android/sdk
    • ANDROID_SDK_ROOT = /Users/lukepighetti/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.52.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.19.0

[✓] Connected device (2 available)
    • Luke's iPhone (mobile) • 0bc0c4864d8e08aac4456893cded68bf0b6767d5 • ios            • iOS 14.4
    • Chrome (web)           • chrome                                   • web-javascript • Google Chrome 88.0.4324.96
    ! Error: Luke's iPhone is busy: Copying cache files from device. Xcode will continue when Luke's iPhone is finished. (code -10)

• No issues found!
```

**Comment by dnfield** [comment]
The migration is available on the master channel, should be available on dev soon.

This will only be available in the flutter sdk vended version, e.g.

```
  dev_dependencies
    integration_test:
      sdk: flutter
```

Unfortunately, this will not make it into the next stable release of Flutter but should be available in either the next dev release or the one after that.

**Comment by ditman** [comment]
Can this be backported to package:integration_test, as a last `.nullsafety` publish? That way people would be able to use this before the sdk gets to stable next?

**Comment by lukepighetti** [comment]
The velocity to beta channel is not ideal.

**Comment by dnfield** [comment]
/cc @stuartmorgan since this is relevant to what we were talking about this morning.

integration_test long term is being moved to the SDK because it has dependencies that come from the SDK (flutter_driver) or need to be exactly in sync with the SDK (vm_service). We can't really publish a null safe version to pub and get anything better than what we already have because it would still have a Flutter SDK version constraint matching at best the commit hash where flutter_driver was migrated, which has missed the boat for the next stable at this point.

We have avoided considering cherry picking integration_test because it would be a fairly large and complicated cherrypick, potentially with impact beyond the package itself (it requires bringing in some package rolls forward that are used elsewhere in the framework and tool).

The impact of this should be that you can migrate your app to null-safety, but you will not be able to migrate anything directly using integration_test to null safety - i.e. you'll have some test files that are not null safe. Unless you use the dev channel that introduces integration_test with NNBD.

I agree it's unfortunate that this is behind other null-safety features we're releaseing for this release. The main reason for that was that certain dependencies of this package took longer to migrate and were deemed as test/tool only and thus given lower priority than packages impacting the framework. There's a meeting coming up with @mit-mit and a few others to discuss this further, and maybe there we can come up with some other solution. I strongly suspect that your best bet will be to track the dev channel until this lands in beta/stable though.

**Comment by dnfield** [comment]
FWIW, `dev` is pretty well vetted. It is used by a lot of production apps, and we tend to catch big things before they land on dev (or get them fixed with pretty high priority if they do manage to land on dev).

**Comment by ditman** [comment]
> We can't really publish a null safe version to pub and get anything better than what we already have because it would still have a Flutter SDK version constraint matching at best the commit hash where flutter_driver was migrated, which has missed the boat for the next stable at this point.

Makes sense, thanks for the thorough explanation!

**Comment by stuartmorgan** [comment]
> I strongly suspect that your best bet will be to track the dev channel until this lands in beta/stable though.

Things become more complicated for plugins, unfortunately, where tracking dev in order to migrate tests affect clients of the plugins.

**Comment by dnfield** [comment]
@jiahaog was going to look at removing the driver dependency. That could change things and make pub work again. But we need to evaluate how that impacts web in particular.

**Comment by lukepighetti** [comment]
I didn't want to kick up any more dust in this thread, but we are also not willing to track `dev` for our packages, only `beta`.

**Comment by MaikuB** [comment]
> Unfortunately, this will not make it into the next stable release of Flutter but should be available in either the next dev release or the one after that.

Given the announcement at https://medium.com/dartlang/preparing-the-dart-and-flutter-ecosystem-for-null-safety-e550ce72c010 that is asking package authors to publish their null safety versions to stable, is this still the case?

I recently filed an issue on the `device_info` plugin at #76194 where it looks like the null safety migration isn't complete. Example app wasn't opted into null safety. This would've caught the issue where null values are sent from Android/iOS via the platform channel and the Dart code has been written to expect non-null values. What concerns me is after doing that, switching to dev channel and updating the `integration_test` to reference the SDK, the integration tests via flutter driver still passed

**Comment by stuartmorgan** [comment]
> Example app wasn't opted into null safety. This would've caught [...]

There's a sweep that just started recently over flutter/plugins to make sure all the `example`s are migrated and run in strong mode. Migrating the example apps themselves isn't actually tied to `integration_test` support.

**Comment by MaikuB** [comment]
@stuartmorgan yes it is indeed but what I forgot to add was that even after migrating the example app, I noticed that integration test wouldn't crash due to issues in the plugin. I was hoping the integration test would catch the issue with the plugin but perhaps I've missed something here on how to get there integration test run in strong mode

**Comment by jlubeck** [comment]
@dnfield what's the proper way to use `integration_tests` now that it's part of the sdk? 
Do we still need to add it inside `dev_dependencies`?

As that is still importing the non null-safety version `1.0.2+2` 

Thank you!

**Comment by stuartmorgan** [comment]
> @dnfield what's the proper way to use `integration_tests` now that it's part of the sdk?

See https://github.com/flutter/flutter/issues/71379#issuecomment-772055489 for the reference format.

**Comment by jlubeck** [comment]
Awesome thank you @stuartmorgan that worked and happy to confirm that the current `dev` channel has this available.

**Comment by bradyt** [comment]
@stuartmorgan Would you know the best way to track when integration_test and/or flutter_driver would be available on beta? Is it reasonable to have an issue dedicated to that?

**Comment by stuartmorgan** [comment]
> @stuartmorgan Would you know the best way to track when integration_test and/or flutter_driver would be available on beta?

When a 2.1.0-* version is released to beta.

> Is it reasonable to have an issue dedicated to that?

Issues reflect the state of master, not release channels, so no.

> Is #77459 likely to get a ping when they're on beta?

There is no mechanism for automatically pinging old bugs when their fixes reach other channels. Someone in the community is welcome to do so if they like.

**Comment by fzyzcjy** [comment]
Tests and NNBD are both very important. So we need to have tests as well as nnbd (and a stable channel - since the app is in production). Is there any workaround, or we have to wait 3 months until the next stable release?

Thanks

**Comment by fzyzcjy** [comment]
In my case, the problem is not simply "the test code needs to be non-null-safe", but my **whole app cannot work!** :(

details: https://github.com/flutter/flutter/issues/82419

**Comment by theochampion** [comment]
Any updates on this? This is blocking

**Comment by jmagman** [comment]
This is available on the [beta channel](https://github.com/flutter/flutter/wiki/Flutter-build-release-channels#how-to-change-channels) and will be in the next stable release.

**Comment by enri90** [comment]
Error: Cannot run with sound null safety, because the following dependencies
don't support null safety:

 - package:integration_test

For solutions, see https://dart.dev/go/unsound-null-safety

:( :(

**Comment by dnfield** [comment]
Nullsafe integration_test is available on the stable channel - please upgrade if you're still seeing this.
